cmd/snap-preseed/preseed_linux.go

```diff
@@ -0,0 +1,172 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/seed"
+	"github.com/snapcore/snapd/timings"
+)
+
+var (
+	// mountPath is where target core/snapd is going to be mounted in the target chroot
+	mountPath     = "/snapd-preseed"
+	syscallMount  = syscall.Mount
+	syscallChroot = syscall.Chroot
+)
+
+// checkChroot does a basic sanity check of the target chroot environment, e.g. makes
+// sure critical virtual filesystems (such as proc) are mounted. This is not meant to
+// be exhaustive check, but one that prevents running the tool against a wrong directory
+// by an accident, which would lead to hard to understand errors from snapd in preseed
+// mode.
+func checkChroot(preseedChroot string) error {
+	exists, isDir, err := osutil.DirExists(preseedChroot)
+	if err != nil {
+		return fmt.Errorf("cannot verify target chroot directory %s: %v", preseedChroot, err)
+	}
+	if !exists || !isDir {
+		return fmt.Errorf("target chroot directory %s doesn't exist or is not a directory", preseedChroot)
+	}
+
+	// sanity checks of the critical mountpoints inside chroot directory
+	for _, p := range []string{"/sys/kernel/security/apparmor", "/proc/self", "/dev/mem"} {
+		path := filepath.Join(preseedChroot, p)
+		if exists := osutil.FileExists(path); !exists {
+			return fmt.Errorf("target chroot directory validation error: %s doesn't exist", path)
+		}
+	}
+
+	return nil
+}
+
+var systemSnapFromSeeds = func(rootDir string) (string, error) {
+	seedDir := filepath.Join(dirs.SnapSeedDirUnder(rootDir))
+	seed, err := seed.Open(seedDir, "")
+	if err != nil {
+		return "", err
+	}
+
+	if err := seed.LoadAssertions(nil, nil); err != nil {
+		return "", err
+	}
+
+	tm := timings.New(nil)
+	if err := seed.LoadMeta(tm); err != nil {
+		return "", err
+	}
+
+	var coreSnapPath string
+	ess := seed.EssentialSnaps()
+	// TODO: handle core18, snapd snap.
+	for _, snap := range ess {
+		if snap.SnapName() == "core" {
+			coreSnapPath = snap.Path
+			break
+		}
+	}
+	if coreSnapPath == "" {
+		return "", fmt.Errorf("core snap not found")
+	}
+
+	return coreSnapPath, nil
+}
+
+func prepareChroot(preseedChroot string) (func(), error) {
+	if err := syscallChroot(preseedChroot); err != nil {
+		return nil, fmt.Errorf("cannot chroot into %s: %v", preseedChroot, err)
+	}
+
+	// GlobalRootDir is now relative to chroot env
+	rootDir := dirs.GlobalRootDir
```

Comment:
After os.Chdir(), we should be able to do `dirs.SetRootDir("/")`, from this point on, dirs.SnapSeedDir should be set correctly.

Reply:
Done. FWTW it was following `image.go` with `dirs.SnapSeedDirUnder(rootDir)`; setting root is not critical since we are only processing seed directory, but I've no strong opinion.

Reply:
I find the SetRootDir a bit confusing personally,  we don't set it before to something else? the point of a  chroot is indeed to change / at which both dirs package values are correct

Reply:
I've reverted this change.